src/compile-gate.lisp

```diff
@@ -230,25 +321,46 @@ This function will compile new ones on-demand."
               (compile-lambda
                (generate-gate-application-code qubits))))))
 
+(defun find-or-make-projector-operator-function (qubit)
+  "Find a projector function for the Hilbert subspace designated by QUBIT (a QUIL:QUBIT).
+
+This function will compile new ones on-demand."
+  (check-type qubit quil:qubit)
+  (let* ((index (quil:qubit-index qubit))
+         (key (list ':measure index)))
+    (or (gethash key **apply-matrix-operator-functions**)
+        (setf (gethash key **apply-matrix-operator-functions**)
+              (compile-lambda
+               (generate-single-qubit-measurement-code index))))))
+
 ;; This function gets called at compile-time in apply-gate.lisp.
 (defun warm-apply-matrix-operator-cache (&key (max-qubits 36))
   "Warm up the **APPLY-MATRIX-OPERATOR-FUNCTIONS** cache for Hilbert spaces B_i and B_i (x) B_j for 0 <= i, j <= MAX-QUBITS."
   (check-type max-qubits nat-tuple-cardinality)
-  ;; Warm the 1q cache.
+  ;; Warm the 1q & measure cache.
   (loop :for q :to max-qubits :do
-    (find-or-make-apply-matrix-operator-function (nat-tuple q)))
+    (find-or-make-apply-matrix-operator-function (nat-tuple q))
+    (find-or-make-projector-operator-function (quil:qubit q)))
   ;; Warm the 2q cache.
   (loop :for p :to max-qubits :do
     (loop :for q :to max-qubits :do
       (unless (= p q)
         (find-or-make-apply-matrix-operator-function (nat-tuple p q)))))
   nil)
 
-(defclass compiled-gate-application (quil:gate-application)
+(defclass compiled-instruction ()       ; Explicitly do *NOT* subclass
+                                        ; from QUIL:INSTRUCTION.
   ((source-instruction :initarg :source-instruction
                        :accessor source-instruction
-                       :documentation "The instruction that was compiled to produce this one.")
-   (source-gate :initarg :source-gate
+                       :documentation "The instruction that was compiled to produce this one."))
+(:metaclass abstract-class))
+
+(defmethod quil::print-instruction-generic ((instr compiled-instruction) stream)
+  (format stream "compiled{ ~/quil:instruction-fmt/ }" (source-instruction instr))
+  nil)
```

Comment:
Does this actually need to return NIL or would `(values)` make sense?

Reply:
either or